### PR TITLE
easy fix for get/volume get/panning in track.py

### PR DESCRIPTION
**Pull request by marcobn**
Hi Daniel,
as promised...
Thanks,
marco

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -127,6 +127,10 @@ See **Device API** for details on type/class_name.
 | /live/clip/set/name | track_id, clip_id, name | | Set clip name |
 | /live/clip/get/gain | track_id, clip_id | gain | Get clip gain |
 | /live/clip/set/gain | track_id, clip_id, gain | | Set clip gain |
+| /live/clip/get/pitch_coarse | track_id, clip_id | semitones | Get clip coarse re-pitch |
+| /live/clip/set/itch_coarse | track_id, clip_id, semitones | | Set clip coarse re-pitch |
```
Comment (ideoforms) [inline README.md:131]: Should be `pitch_coarse`

File: abletonosc/track.py [hunk 2]
```diff
@@ -39,13 +39,15 @@ def track_callback(params: Tuple[Any]):
                                         create_track_callback(self._set, prop))
 
         def track_get_volume(track, params: Tuple[Any] = ()):
-            return track.mixer_device.volume.value
+            params = (track.mixer_device.volume.value,)
+            return params
```
Comment (ideoforms) [inline abletonosc/track.py:43]: Silly mistake on my part - clearly this callback has never worked! 
For consistency, could you switch this to the same syntax used in (e.g.) song.py:

```
        def song_get_num_tracks(song, params: Tuple[Any] = ()):
            return len(song.tracks),
```

Saves a line of code and the creation of a temporary `params` tuple.

**Review by ideoforms — CHANGES_REQUESTED** [review]
Thanks for the PR. Two small amends, and then I'll merge this tomorrow. 
I will also start work on some `track` unit tests so that bugs like this don't slip through...

**Review by ideoforms — APPROVED** [review]

**Comment by ideoforms** [comment]
Thanks!

**Comment by marcobn** [comment]
Hi Daniel,
I have another question for you… I noticed that the return values for volume in track or gain in clips is apparently on a linear scale between 0 and 1. Now, the levels in Live are of course on a dB log scale of some sort (-inf to 6 in tracks and -inf to 20 in clips). In this way, 0 dB in tracks returns 0.85, while 0 dB in clips returns 0.4. Any insight?
Thanks,
marco


Marco Buongiorno Nardelli (he/him)
Regents Professor, University of North Texas
External Professor, Santa Fe Institute
CEMI, Center for Experimental Music and Intermedia
iARTA, Initiative for Advanced Research in Technology and the Arts
ArtSciLab, ATEC @ the University of Texas at Dallas
http://ermes.unt.edu
http://www.materialssoundmusic.com



From: Daniel Jones ***@***.***>
Date: Sunday, October 9, 2022 at 3:51 AM
To: ideoforms/AbletonOSC ***@***.***>
Cc: Buongiorno Nardelli, Marco ***@***.***>, Author ***@***.***>
Subject: [EXT] Re: [ideoforms/AbletonOSC] easy fix for get/volume get/panning in track.py (PR #15)

Thanks!

**Comment by ideoforms** [comment]
Hmm, it's a good question. Looking at the track-level volume sliders in Live, it looks like it ranges from -70dB to +6dB. But, the negative portion of this (in dB) isn't linearly mapped to the 0..1 float scale, so it is not as simple as a linear/logarithmic scaling - there's some additional curve going on. I don't see anything in the [LOM docs](https://docs.cycling74.com/max8/vignettes/live_object_model) that immediately jumps out as helpful.

What are you trying to achieve?

**Comment by jln-** [comment]
Hi. If it helps, there's a Max abstraction call M4L.api.db2value that implements the dB to Live parameter linear scale.
Capture d’écran 2022-10-09 à 23 13 04

**Comment by marcobn** [comment]
Here is a pythonic way to convert Live  linear values to dB.

For track volume:

def db2value(db):
    # conversion from decibel to linear scale in track volume
    if db <= 6 and db >= -18:
        return (db+34)/40
    elif db < -18 and db >= -41:
        alpha = 799.503788
        beta = 12630.61132
        gamma = 201.871345
        delta = 399.751894
        return -(np.sqrt(-alpha*db - beta) - gamma) / delta
    elif db < -41:
        alpha = 70.
        beta = 118.426374
        gamma = 7504./5567.
        return np.power(((db+alpha)/beta),gamma)
    else:
        print('out of bounds')

and its reciprocal:

def value2db(vl):
    # conversion from linear to decibel scale in track volume
    if vl <= 1 and vl >= 0.4:
        return 40*vl -34
    elif vl < 0.4 and vl >= 0.15:
        alpha = 799.503788
        beta = 12630.61132
        gamma = 201.871345
        delta = 399.751894
        return -((delta*vl - gamma)**2 + beta)/alpha
    elif vl < 0.15:
        alpha = 70.
        beta = 118.426374
        gamma = 7504./5567.
        return beta*np.power(vl,1/gamma) - alpha
    else:
        print('out of bounds')

For gain in clips

def db4value(db):
    # conversion from decibel to linear scale in clip gain
    db -= 18
    if db <= 6 and db >= -18:
        return (db+34)/40
    elif db < -18 and db >= -41:
        alpha = 799.503788
        beta = 12630.61132
        gamma = 201.871345
        delta = 399.751894
        return -(np.sqrt(-alpha*db - beta) - gamma) / delta
    elif db < -41:
        alpha = 70.
        beta = 118.426374
        gamma = 7504./5567.
        return np.power(((db+alpha)/beta),gamma)
    else:
        print('out of bounds')

for the reciprocal, one can retrieve /live/clip/get/gain_display_string. It will return the value in dB.
marco

Marco Buongiorno Nardelli (he/him)
Regents Professor, University of North Texas
External Professor, Santa Fe Institute
CEMI, Center for Experimental Music and Intermedia
iARTA, Initiative for Advanced Research in Technology and the Arts
ArtSciLab, ATEC @ the University of Texas at Dallas
http://ermes.unt.edu
http://www.materialssoundmusic.com



From: Daniel Jones ***@***.***>
Date: Sunday, October 9, 2022 at 1:42 PM
To: ideoforms/AbletonOSC ***@***.***>
Cc: Buongiorno Nardelli, Marco ***@***.***>, Author ***@***.***>
Subject: [EXT] Re: [ideoforms/AbletonOSC] easy fix for get/volume get/panning in track.py (PR #15)

Hmm, it's a good question. Looking at the track-level volume sliders in Live, it looks like it ranges from -70dB to +6dB. But, the negative portion of this (in dB) isn't linearly mapped to the 0..1 float scale, so it is not as simple as a linear/logarithmic scaling - there's some additional curve going on. I don't see anything in the LOM docs<https://nam04.safelinks.protection.outlook.com/?url=https%3A%2F%2Fdocs.cycling74.com%2Fmax8%2Fvignettes%2Flive_object_model&data=05%7C01%7CMarco.BuongiornoNardelli%40unt.edu%7Ce6eec04a9974480d2bcc08daaa2611fc%7C70de199207c6480fa318a1afcba03983%7C0%7C0%7C638009377755561808%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=EPN9iBj1Zx234qaaquGIazouTkpER%2Fh%2FXoOj6Ok0Zx8%3D&reserved=0> that immediately jumps out as helpful.

What are you trying to achieve?

**Pull request merged**